File: 000Unordered_ Videos/WWDC 2013/WWDC 2013 Sample Code/Folder/TicTacToeApp/TicTacToe/TTTGameView.m
/*
     File: TTTGameView.m
 Abstract: 
 
  Version: 1.0
 
 Disclaimer: IMPORTANT:  This Apple software is supplied to you by Apple
 Inc. ("Apple") in consideration of your agreement to the following
 terms, and your use, installation, modification or redistribution of
 this Apple software constitutes acceptance of these terms.  If you do
 not agree with these terms, please do not use, install, modify or
 redistribute this Apple software.
 
 In consideration of your agreement to abide by the following terms, and
 subject to these terms, Apple grants you a personal, non-exclusive
 license, under Apple's copyrights in this original Apple software (the
 "Apple Software"), to use, reproduce, modify and redistribute the Apple
 Software, with or without modifications, in source and/or binary forms;
 provided that if you redistribute the Apple Software in its entirety and
 without modifications, you must retain this notice and the following
 text and disclaimers in all such redistributions of the Apple Software.
 Neither the name, trademarks, service marks or logos of Apple Inc. may
 be used to endorse or promote products derived from the Apple Software
 without specific prior written permission from Apple.  Except as
 expressly stated in this notice, no other rights or licenses, express or
 implied, are granted by Apple herein, including but not limited to any
 patent rights that may be infringed by your derivative works or by other
 works in which the Apple Software may be incorporated.
 
 The Apple Software is provided by Apple on an "AS IS" basis.  APPLE
 MAKES NO WARRANTIES, EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION
 THE IMPLIED WARRANTIES OF NON-INFRINGEMENT, MERCHANTABILITY AND FITNESS
 FOR A PARTICULAR PURPOSE, REGARDING THE APPLE SOFTWARE OR ITS USE AND
 OPERATION ALONE OR IN COMBINATION WITH YOUR PRODUCTS.
 
 IN NO EVENT SHALL APPLE BE LIABLE FOR ANY SPECIAL, INDIRECT, INCIDENTAL
 OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 INTERRUPTION) ARISING IN ANY WAY OUT OF THE USE, REPRODUCTION,
 MODIFICATION AND/OR DISTRIBUTION OF THE APPLE SOFTWARE, HOWEVER CAUSED
 AND WHETHER UNDER THEORY OF CONTRACT, TORT (INCLUDING NEGLIGENCE),
 STRICT LIABILITY OR OTHERWISE, EVEN IF APPLE HAS BEEN ADVISED OF THE
 POSSIBILITY OF SUCH DAMAGE.
 
 Copyright (C) 2013 Apple Inc. All Rights Reserved.
 
 
 Copyright © 2013 Apple Inc. All rights reserved.
 WWDC 2013 License
 
 NOTE: This Apple Software was supplied by Apple as part of a WWDC 2013
 Session. Please refer to the applicable WWDC 2013 Session for further
 information.
 
 IMPORTANT: This Apple software is supplied to you by Apple Inc.
 ("Apple") in consideration of your agreement to the following terms, and
 your use, installation, modification or redistribution of this Apple
 software constitutes acceptance of these terms. If you do not agree with
 these terms, please do not use, install, modify or redistribute this
 Apple software.
 
 In consideration of your agreement to abide by the following terms, and
 subject to these terms, Apple grants you a non-exclusive license, under
 Apple's copyrights in this original Apple software (the "Apple
 Software"), to use, reproduce, modify and redistribute the Apple
 Software, with or without modifications, in source and/or binary forms;
 provided that if you redistribute the Apple Software in its entirety and
 without modifications, you must retain this notice and the following
 text and disclaimers in all such redistributions of the Apple Software.
 Neither the name, trademarks, service marks or logos of Apple Inc. may
 be used to endorse or promote products derived from the Apple Software
 without specific prior written permission from Apple. Except as
 expressly stated in this notice, no other rights or licenses, express or
 implied, are granted by Apple herein, including but not limited to any
 patent rights that may be infringed by your derivative works or by other
 works in which the Apple Software may be incorporated.
 
 The Apple Software is provided by Apple on an "AS IS" basis. APPLE MAKES
 NO WARRANTIES, EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION THE
 IMPLIED WARRANTIES OF NON-INFRINGEMENT, MERCHANTABILITY AND FITNESS FOR
 A PARTICULAR PURPOSE, REGARDING THE APPLE SOFTWARE OR ITS USE AND
 OPERATION ALONE OR IN COMBINATION WITH YOUR PRODUCTS.
 
 IN NO EVENT SHALL APPLE BE LIABLE FOR ANY SPECIAL, INDIRECT, INCIDENTAL
 OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 INTERRUPTION) ARISING IN ANY WAY OUT OF THE USE, REPRODUCTION,
 MODIFICATION AND/OR DISTRIBUTION OF THE APPLE SOFTWARE, HOWEVER CAUSED
 AND WHETHER UNDER THEORY OF CONTRACT, TORT (INCLUDING NEGLIGENCE),
 STRICT LIABILITY OR OTHERWISE, EVEN IF APPLE HAS BEEN ADVISED OF THE
 POSSIBILITY OF SUCH DAMAGE.
 
 EA1002
 5/3/2013
 */

#import "TTTGameView.h"

#import "TTTGame.h"

@interface TTTGameLineView : UIView
@property (copy, nonatomic) UIBezierPath *path;
@property (strong, nonatomic) UIColor *color;
@end

@implementation TTTGameView {
    NSArray *_horizontalLineViews;
    NSArray *_verticalLineViews;
    NSMutableArray *_moveImageViews;
    NSMutableArray *_moveImageViewReuseQueue;
    TTTGameLineView *_lineView;
}

- (id)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if (self) {
        _gridColor = [UIColor blackColor];
        _horizontalLineViews = @[[self lineView], [self lineView]];
        _verticalLineViews = @[[self lineView], [self lineView]];
        [self updateGridColor];
        
        _moveImageViews = [[NSMutableArray alloc] init];
        _moveImageViewReuseQueue = [[NSMutableArray alloc] init];
        
        UITapGestureRecognizer *gestureRecognizer = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(tapGame:)];
        [self addGestureRecognizer:gestureRecognizer];
    }
    return self;
}

- (UIView *)lineView
{
    UIView *view = [[UIView alloc] init];
    [self addSubview:view];
    return view;
}

- (void)tapGame:(UITapGestureRecognizer *)gestureRecognizer
{
    if ((gestureRecognizer.state == UIGestureRecognizerStateRecognized) && [self.delegate respondsToSelector:@selector(gameView:didSelectXPosition:yPosition:)]) {
        CGPoint point = [gestureRecognizer locationInView:self];
        CGRect bounds = self.bounds;
        
        CGPoint normalizedPoint = point;
        normalizedPoint.x -= CGRectGetMidX(bounds);
        normalizedPoint.x *= 3.0 / bounds.size.width;
        normalizedPoint.x = round(normalizedPoint.x);
        normalizedPoint.x = MAX(normalizedPoint.x, -1);
        normalizedPoint.x = MIN(normalizedPoint.x, 1);
        TTTMoveXPosition xPosition = normalizedPoint.x;
        
        normalizedPoint.y -= CGRectGetMidY(bounds);
        normalizedPoint.y *= 3.0 / bounds.size.height;
        normalizedPoint.y = round(normalizedPoint.y);
        normalizedPoint.y = MAX(normalizedPoint.y, -1);
        normalizedPoint.y = MIN(normalizedPoint.y, 1);
        TTTMoveYPosition yPosition = normalizedPoint.y;
        
        if (![self.delegate respondsToSelector:@selector(gameView:canSelectXPosition:yPosition:)] || [self.delegate gameView:self canSelectXPosition:xPosition yPosition:yPosition]) {
            [self.delegate gameView:self didSelectXPosition:xPosition yPosition:yPosition];
        }
    }
}

- (void)setGame:(TTTGame *)value
{
    if (_game != value) {
        _game = value;
        [self updateGameState];
    }
}

- (UIImageView *)moveImageView
{
    UIImageView *moveView = [_moveImageViewReuseQueue firstObject];
    if (moveView) {
        [_moveImageViewReuseQueue removeObject:moveView];
    } else {
        moveView = [[UIImageView alloc] init];
        [self addSubview:moveView];
    }
    [_moveImageViews addObject:moveView];
    return moveView;
}

- (CGPoint)pointForXPosition:(TTTMoveXPosition)xPosition yPosition:(TTTMoveYPosition)yPosition
{
    CGRect bounds = self.bounds;
    CGPoint point = CGPointMake(CGRectGetMidX(bounds), CGRectGetMidY(bounds));
    point.x += xPosition * bounds.size.width / 3.0;
    point.y += yPosition * bounds.size.height / 3.0;
    return point;
}

- (void)setMove:(TTTMove *)move forView:(UIImageView *)moveView
{
    moveView.image = [self.delegate gameView:self imageForPlayer:move.player];
    moveView.center = [self pointForXPosition:move.xPosition yPosition:move.yPosition];
}

- (void)setVisible:(BOOL)visible forView:(UIImageView *)moveView
{
    if (visible) {
        [moveView sizeToFit];
        moveView.alpha = 1.0;
    } else {
        moveView.bounds = CGRectZero;
        moveView.alpha = 0.0;
    }
}

- (void)updateGameState
{
    NSArray *moves = self.game.moves;
    NSInteger moveCount = moves.count;
    [[_moveImageViews copy] enumerateObjectsUsingBlock:^(UIImageView *moveView, NSUInteger viewIndex, BOOL *stop) {
        if (viewIndex < moveCount) {
            TTTMove *move = moves[viewIndex];
            [self setMove:move forView:moveView];
            [self setVisible:YES forView:moveView];
        } else {
            [self setVisible:NO forView:moveView];
            [_moveImageViewReuseQueue addObject:moveView];
            [_moveImageViews removeObject:moveView];
        }
    }];
    
    for (NSInteger moveIndex = _moveImageViews.count; moveIndex < moveCount; moveIndex++) {
        TTTMove *move = moves[moveIndex];
        UIImageView *moveView = [self moveImageView];
        [UIView performWithoutAnimation:^{
            [self setMove:move forView:moveView];
            [self setVisible:NO forView:moveView];
        }];
        
        [self setVisible:YES forView:moveView];
    }
    
    TTTMovePlayer winningPlayer;
    TTTMoveXPosition startXPosition, endXPosition;
    TTTMoveYPosition startYPosition, endYPosition;
    BOOL hasWinner = [self.game getWinningPlayer:&winningPlayer startXPosition:&startXPosition startYPosition:&startYPosition endXPosition:&endXPosition endYPosition:&endYPosition];
    if (hasWinner) {
        if (!_lineView) {
            _lineView = [[TTTGameLineView alloc] init];
            _lineView.alpha = 0.0;
            [self addSubview:_lineView];
        }
        UIBezierPath *path = [UIBezierPath bezierPath];
        [path moveToPoint:[self pointForXPosition:startXPosition yPosition:startYPosition]];
        [path addLineToPoint:[self pointForXPosition:endXPosition yPosition:endYPosition]];
        _lineView.path = path;
        _lineView.color = [self.delegate gameView:self colorForPlayer:winningPlayer];
    }
    _lineView.alpha = (hasWinner ? 1.0 : 0.0);
}

static CGFloat const TTTGameViewLineWidth = 4.0;

- (void)layoutSubviews
{
    [super layoutSubviews];
    
    CGRect bounds = self.bounds;
    [_horizontalLineViews enumerateObjectsUsingBlock:^(UIView *view, NSUInteger viewIndex, BOOL *stop) {
        view.bounds = CGRectMake(0.0, 0.0, bounds.size.width, TTTGameViewLineWidth);
        view.center = CGPointMake(CGRectGetMidX(bounds), round(bounds.size.height * (viewIndex + 1) / 3.0));
    }];
    [_verticalLineViews enumerateObjectsUsingBlock:^(UIView *view, NSUInteger viewIndex, BOOL *stop) {
        view.bounds = CGRectMake(0.0, 0.0, TTTGameViewLineWidth, bounds.size.height);
        view.center = CGPointMake(round(bounds.size.width * (viewIndex + 1) / 3.0), CGRectGetMidY(bounds));
    }];
    [self updateGameState];
}

- (void)setGridColor:(UIColor *)value
{
    if (![_gridColor isEqual:value]) {
        _gridColor = value;
        [self updateGridColor];
    }
}

- (void)updateGridColor
{
    for (UIView *view in _horizontalLineViews) {
        view.backgroundColor = self.gridColor;
    }
    for (UIView *view in _verticalLineViews) {
        view.backgroundColor = self.gridColor;
    }
}

@end

@implementation TTTGameLineView

+ (Class)layerClass
{
    return [CAShapeLayer class];
}

- (CAShapeLayer *)shapeLayer
{
    return (CAShapeLayer *)[self layer];
}

- (id)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if (self) {
        self.shapeLayer.lineWidth = 2.0;
    }
    return self;
}

- (void)setPath:(UIBezierPath *)value
{
    if (_path != value) {
        _path = [value copy];
        self.shapeLayer.path = [_path CGPath];
    }
}

- (void)setColor:(UIColor *)value
{
    if (![_color isEqual:value]) {
        _color = value;
        self.shapeLayer.strokeColor = [_color CGColor];
    }
}

@end

File: UIKit and Animations/WWDC 2013/Customizing Your App's Appearance for iOS 7/Sample_Code/TicTacToeApp_Swift/TicTacToeApp_Swift/View/TTTGameView.swift
//
//  TTTGameView.swift
//  TicTacToeApp_Swift
//
//  Created by Seyed Samad Gholamzadeh on 7/12/18.
//  Copyright © 2018 Seyed Samad Gholamzadeh. All rights reserved.
//

import UIKit

let TTTGameViewLineWidth: CGFloat = 4.0

protocol TTTGameViewDelegate: class {
	
	func gameView(_ gameView: TTTGameView, imageFor player: TTTMovePlayer) -> UIImage
	func gameView(_ gameView: TTTGameView, colorFor player: TTTMovePlayer) -> UIColor
	
	func gameView(_ gameView: TTTGameView, canSelect xPosition: TTTMoveXPosition, yPosition:TTTMoveYPosition) -> Bool
	func gameView(_ gameView: TTTGameView, didSelect xPosition: TTTMoveXPosition, yPosition:TTTMoveYPosition)
}

extension TTTGameViewDelegate {
	func gameView(_ gameView: TTTGameView, canSelect xPosition: TTTMoveXPosition, yPosition:TTTMoveYPosition) -> Bool {
		return false
	}
	func gameView(_ gameView: TTTGameView, didSelect xPosition: TTTMoveXPosition, yPosition:TTTMoveYPosition) {
		
	}

}

class TTTGameView: UIView {

	var delegate: TTTGameViewDelegate?
	var game: TTTGame! {
		didSet {
			if self.game != nil {
				self.updateGameState()
			}
		}
	}
	
	var gridColor: UIColor! {
		didSet {
			if self.gridColor != nil {
				self.updateGridColor()
			}
		}
	}
	var horizontalLineViews: [UIView]!
	var verticalLineViews: [UIView]!
	var moveImageViews: [UIImageView]!
	var moveImageViewReuseQueue: [UIImageView]!
	var _lineView: TTTGameLineView!
	
	var moveImageView: UIImageView {
		var moveView = self.moveImageViewReuseQueue.first
		if moveView != nil {
			self.moveImageViewReuseQueue.remove(at: self.moveImageViewReuseQueue.index(of: moveView!)!)
		} else {
			moveView = UIImageView()
			self.addSubview(moveView!)
		}
		
		self.moveImageViews.append(moveView!)
		return moveView!
	}
	
	override init(frame: CGRect) {
		super.init(frame: frame)
		self.gridColor = UIColor.black
		self.horizontalLineViews =  [self.lineView(), self.lineView()]
		self.verticalLineViews =  [self.lineView(), self.lineView()]
		self.updateGridColor()
		
		self.moveImageViews = []
		self.moveImageViewReuseQueue = []
		
		let gestureRecognizer = UITapGestureRecognizer(target: self, action: #selector(tapGame(_:)))
		self.addGestureRecognizer(gestureRecognizer)
	}
	
	required init?(coder aDecoder: NSCoder) {
		fatalError("init(coder:) has not been implemented")
	}
	
	func lineView() -> UIView {
		let view = UIView()
		self.addSubview(view)
		return view
	}
	
	@objc func tapGame(_ gestureRecognizer: UITapGestureRecognizer) {
		if gestureRecognizer.state == .recognized {
			let point = gestureRecognizer.location(in: self)
			let bounds = self.bounds
			
			var normalizedPoint = point
			normalizedPoint.x -= bounds.midX
			normalizedPoint.x *= 3.0 / bounds.size.width
			normalizedPoint.x = round(normalizedPoint.x)
			normalizedPoint.x = max(normalizedPoint.x, -1)
			normalizedPoint.x = min(normalizedPoint.x, 1)
			let xPosition = normalizedPoint.x;
			
			normalizedPoint.y -= bounds.midY
			normalizedPoint.y *= 3.0 / bounds.size.height
			normalizedPoint.y = round(normalizedPoint.y)
			normalizedPoint.y = max(normalizedPoint.y, -1)
			normalizedPoint.y = min(normalizedPoint.y, 1)
			let yPosition = normalizedPoint.y
			
			if (self.delegate?.gameView(self, canSelect: TTTMoveXPosition(rawValue: Int(xPosition))!, yPosition: TTTMoveYPosition(rawValue: Int(yPosition))!))! {
				self.delegate?.gameView(self, didSelect: TTTMoveXPosition(rawValue: Int(xPosition))!, yPosition: TTTMoveYPosition(rawValue: Int(yPosition))!)
			}

		}
	}
	
	func point(for xPosition: TTTMoveXPosition, yPosition: TTTMoveYPosition) -> CGPoint {
		let bounds = self.bounds;
		var point = CGPoint(x: bounds.midX, y: bounds.midY)
		point.x += CGFloat(xPosition.rawValue) * bounds.size.width / 3.0
		point.y += CGFloat(yPosition.rawValue) * bounds.size.height / 3.0
		return point
	}
	
	func set(_ move: TTTMove, for moveView: UIImageView) {
		moveView.image = self.delegate?.gameView(self, imageFor: move.player)
		moveView.center = self.point(for: move.xPosition, yPosition: move.yPosition)
	}
	
	func setVisible(_ visible: Bool, for moveView: UIImageView) {
		if (visible) {
			moveView.sizeToFit()
			moveView.alpha = 1.0
		} else {
			moveView.bounds = CGRect.zero
			moveView.alpha = 0.0
		}
	}
	
	func updateGameState() {
		let moves = self.game?.moves ?? []
		let moveCount = moves.count
		let moveImageViews = self.moveImageViews ?? []
		
		for (viewIndex, moveView) in moveImageViews.enumerated() {
			if (viewIndex < moveCount) {
				let move = moves[viewIndex]
				self.set(move, for: moveView)
				self.setVisible(true, for: moveView)
			} else {
				self.setVisible(false, for: moveView)
				self.moveImageViewReuseQueue.append(moveView)
				self.moveImageViews.remove(at: self.moveImageViews.index(of: moveView)!)
			}
		}
		
		for moveIndex in self.moveImageViews.count..<moveCount {
			let move = moves[moveIndex]
			let moveView = self.moveImageView
			UIView.performWithoutAnimation {
				self.set(move, for: moveView)
				self.setVisible(false, for: moveView)
			}
			
			self.setVisible(true, for: moveView)
		}
		
		var winningPlayer: TTTMovePlayer!
		var startXPosition, endXPosition: TTTMoveXPosition!
		var startYPosition, endYPosition: TTTMoveYPosition!
		let hasWinner = self.game.getWinningPlayer(&winningPlayer, startXPosition: &startXPosition, startYPosition: &startYPosition, endXPosition: &endXPosition, endYPosition: &endYPosition)
		if hasWinner {
			if self._lineView == nil {
				self._lineView = TTTGameLineView()
				_lineView.alpha = 0.0
				self.addSubview(self._lineView)
			}
			let path = UIBezierPath()
			path.move(to: self.point(for: startXPosition, yPosition: startYPosition))
			path.addLine(to: self.point(for: endXPosition, yPosition: endYPosition))
			self._lineView.path = path
			self._lineView.color = self.delegate?.gameView(self, colorFor: winningPlayer)
		}
		self._lineView?.alpha = hasWinner ? 1.0 : 0.0
	}
	
	override func layoutSubviews() {
		super.layoutSubviews()
		
		let bounds = self.bounds
		
		for (viewIndex, view) in self.horizontalLineViews.enumerated() {
			view.bounds = CGRect(x: 0, y: 0, width: bounds.size.width, height: TTTGameViewLineWidth)
			view.center = CGPoint(x: bounds.midX, y: round(bounds.size.height * CGFloat(viewIndex + 1) / 3.0))
		}
		
		for (viewIndex, view) in self.verticalLineViews.enumerated() {
			view.bounds = CGRect(x: 0, y: 0, width: TTTGameViewLineWidth, height: bounds.size.height)
			view.center = CGPoint(x: round(bounds.size.width * CGFloat(viewIndex + 1) / 3.0), y: bounds.midY)
		}
		self.updateGameState()
	}

	func updateGridColor() {
		for view in self.horizontalLineViews {
			view.backgroundColor = self.gridColor
		}
		
		for view in self.verticalLineViews {
			view.backgroundColor = self.gridColor
		}
	}
	
}


class TTTGameLineView: UIView {
	var path: UIBezierPath! {
		didSet {
			if self.path != nil {
				self.shapeLayer.path = self.path.cgPath
			}
		}
	}
	
	var color: UIColor! {
		didSet {
			if self.color != nil {
				self.shapeLayer.strokeColor = self.color.cgColor
			}
		}
	}
	
	override class var layerClass: AnyClass {
		return CAShapeLayer.self
	}
	
	var shapeLayer: CAShapeLayer {
		return (self.layer as! CAShapeLayer)
	}
	
	override init(frame: CGRect) {
		super.init(frame: frame)
		self.shapeLayer.lineWidth = 2.0
	}
	
	required init?(coder aDecoder: NSCoder) {
		fatalError("init(coder:) has not been implemented")
	}
	
}
